core/src/snapshot_packager_service.rs

```diff
@@ -42,24 +45,46 @@ impl SnapshotPackagerService {
                     }
 
                     let snapshot_package = pending_snapshot_package.lock().unwrap().take();
-                    if let Some(snapshot_package) = snapshot_package {
-                        match snapshot_utils::archive_snapshot_package(
+                    if snapshot_package.is_none() {
+                        std::thread::sleep(Duration::from_millis(100));
+                        continue;
+                    }
+                    let snapshot_package = snapshot_package.unwrap();
+
+                    let mut attempts = 0;
+                    loop {
+                        attempts += 1;
+                        let result = snapshot_utils::archive_snapshot_package(
                             &snapshot_package,
                             maximum_snapshots_to_retain,
-                        ) {
-                            Ok(_) => {
-                                hashes.push((snapshot_package.slot(), *snapshot_package.hash()));
-                                while hashes.len() > MAX_SNAPSHOT_HASHES {
-                                    hashes.remove(0);
-                                }
-                                cluster_info.push_snapshot_hashes(hashes.clone());
-                            }
-                            Err(err) => {
-                                warn!("Failed to create snapshot archive: {}", err);
+                        );
+
+                        if result.is_ok() {
+                            hashes.push((snapshot_package.slot(), *snapshot_package.hash()));
+                            while hashes.len() > MAX_SNAPSHOT_HASHES {
+                                hashes.remove(0);
                             }
-                        };
-                    } else {
-                        std::thread::sleep(Duration::from_millis(100));
+                            cluster_info.push_snapshot_hashes(hashes.clone());
+                            break;
+                        }
+
+                        warn!(
+                            "Failed to create snapshot archive. attempts: {}, slot: {}, err: {:?}",
+                            attempts,
+                            snapshot_package.slot(),
+                            result
+                        );
+
+                        // Only require success for full snapshots.  Otherwise just log the error and
+                        // continue on.
+                        if snapshot_package.snapshot_type != SnapshotType::FullSnapshot {
+                            break;
+                        }
+
+                        assert!(
+                            attempts <= ARCHIVE_SNAPSHOT_PACKAGE_MAX_ATTEMPTS,
+                            "Unable to archive full snapshot package!"
+                        );
                     }
```

Comment:
Here's the new loop where we retry archiving full snapshot packages if there's a failure. Is `assert`ing the best option here? Will that bring down the whole node, or just this thread?

Reply:
hehe yeah its a puzzle whether the web of channels we have implemented will correctly shut down a node in cases where a specific thread panics. Generally, threads that are connected via channels will detect the other side has closed and also close, but that is not the case here (no channels to any other threads). 

I think you might have to set `exit` to true here and log an error. It's the global validator exit flag, so should shut down the full validator (maybe a good test is to spin up a validator with this exit set to true after 30 seconds)

Reply:
Good idea, will do.

Reply:
OK! I confirmed that the asserts will bring down the whole node. I ran bootstrap-validator locally and inserted `panic!()`s into both AccountsBackgroundService and SnapshotPackagerService separately. The panics fired and immediately killed the process (then the script restarted the validator, as expected, successfully). So looks like this is all good.